source/h1_connection.c

```diff
@@ -743,68 +688,56 @@ static struct aws_h1_stream *s_update_outgoing_stream_ptr(struct h1_connection *
         current = NULL;
     }
 
-    /* If current stream is NULL, do one of:
-     * - Look for next item in stream_list.
-     * - Client side: look in synced_data.pending_stream_list for more work
-     * - Server side: look in thread_data.waiting_stream_list for more work */
+    /* If current stream is NULL, look for more work:
+     * If next_node in stream_list exists, work on that.
+     * Else, look for more work in other lists.
+     *      If server: look in thread_data.waiting_stream_list.
+     *      If client: look in synced_data.pending_stream_list. */
     if (!current && !connection->thread_data.is_writing_stopped) {
         if (next_node && next_node != aws_linked_list_end(&connection->thread_data.stream_list)) {
             current = AWS_CONTAINER_OF(next_node, struct aws_h1_stream, node);
-
-        } else if (connection->base.server_data) {
-            /* server side should check the stream already has response or not
-             * Require a lock to prevent the user makes any change to the stream state */
+        } else /* Look for more work in other lists */ {
             /* BEGIN CRITICAL SECTION */
             s_h1_connection_lock_synced_data(connection);
-            while (!aws_linked_list_empty(&connection->thread_data.waiting_stream_list)) {
-                /* The front of waiting_stream_list is not ready to be sent */
-                if (!AWS_CONTAINER_OF(
-                         aws_linked_list_front(&connection->thread_data.waiting_stream_list),
-                         struct aws_h1_stream,
-                         node)
-                         ->synced_data.has_outgoing_response) {
-                    break;
+
+            if (connection->base.server_data) {
+                /* Move streams from front of waiting_stream_list to back of stream_list
+                 * UNTIL we hit a stream whose response isn't ready */
+                struct aws_linked_list *waiting_stream_list = &connection->thread_data.waiting_stream_list;
```

Comment:
NIT: for future posterity, could you document why we're not swapping the list and immediately releasing the lock?

Reply:
I was trying to say that with this comment here. I'll be more clear